### PR TITLE
fix(llm): stop charging event-analysis spend to the chat budget

### LINKED ISSUE #290: [BUG] - AI chat budget runs out early because event-analysis spend is charged to it
## Description

Nudgebee tracks two separate AI spending budgets per account/organization: one for **AI chat** (user-initiated "Nubi" conversations) and one for **Event Analysis** (automatic investigation of events). These are configured as independent budgets with their own limits.

In practice the **AI chat budget is being consumed by Event Analysis spend as well** — every dollar spent on automatic event analysis is counted against the chat budget *on top of* its own event-analysis budget. As a result the chat budget shows far more usage than users have actually generated by chatting, its "remaining" figure is understated, and chat can be throttled/blocked even when users have barely used it.

## Impact

- **Who is affected**: All tenants/accounts that have AI-chat budget limits enabled and also run Event Analysis (effectively everyone using both features).
- **Severity**: AI chat can hit its budget limit and start returning "budget exceeded" errors prematurely, blocking legitimate user chats. Budget/usage reporting for chat is inflated and misleading.
- **Since when**: Unknown — present since the multi-module budget usage logic was introduced.

## Reproduction Steps

This is a backend calculation issue; it surfaces in the budget status shown for an account. To observe it:

1. Pick an account that has run Event Analysis this month (non-zero event-analysis spend).
2. Look at the budget status for that account (`ai_get_budget_status`).
3. Compare the `investigation` (Event Analysis) monthly cost usage against the `user_investigation` (AI chat) monthly cost usage.
4. Observe that the chat (`user_investigation`) usage is *always >=* the event usage and in fact includes it — chat usage minus event usage is the true chat spend.

Counts make this unmistakable: an account showing, say, 110 event conversations and 20,359 chat conversations — the 110 event conversations are a subset of the 20,359.

---

## Technical Details

Budget usage is computed per "module" in `llm/llm-server/budget`. The module → SQL filter map in `llm/llm-server/budget/types.go` is:

```go
var moduleQueryFilters = map[string]string{
    ModuleInvestigation:     " AND c.session_id LIKE '" + events.SessionIdPrefixEvent + "%'", // 'event-%'
    ModuleUserInvestigation: "", // <-- empty: matches ALL conversations, including event-*
}
```

`getEntityTokenUsage` / `getEntityConversationCount` in `llm/llm-server/budget/usage.go` apply this filter when summing cost from `llm_conversation_token_usage` (joined to `llm_model_pricing`) and counting `llm_conversations`. Because the `user_investigation` filter is empty, its usage/count query sweeps in **every** conversation for the entity — including the `event-` prefixed event-analysis sessions that `investigation` already accounts for. Hence the double-count.

This contradicts the **enforcement** logic, which is mutually exclusive by session prefix (event → `investigation`, else → `user_investigation`):

- `llm/llm-server/api/chains.go` (around lines 405, 784, 1330)
- `llm/llm-server/api/conversation_sync.go` (around line 101)

So enforcement treats the two budgets as disjoint, but usage summation treats `user_investigation` as a superset. The same `getEntityTokenUsage` backs both reporting (`GetBudgetStatus`) and enforcement (`CheckBudgetLimits`), so the inflated number also drives premature `429`s.

## Proposed fix

Make `user_investigation` the complement of `investigation`:

```go
ModuleUserInvestigation: " AND c.session_id NOT LIKE '" + events.SessionIdPrefixEvent + "%'",
```

(`event-rca-` sessions also start with `event-`, so a single `NOT LIKE 'event-%'` excludes them too.)

**Open question before fixing:** confirm that *every* non-event conversation in `llm_conversations` is genuinely an AI-chat session. If other session types exist (scheduled jobs, automations, internal agents), `NOT LIKE 'event-%'` would still lump them into chat. If so, the correct fix is a positive prefix match for chat sessions rather than a negative event match.

## Acceptance

- `user_investigation` usage no longer includes event-analysis spend.
- `GetBudgetStatus` and `CheckBudgetLimits` agree on the partition (matching the prefix-based routing in `chains.go` / `conversation_sync.go`).
- Regression test in `llm/llm-server/budget/service_test.go` asserting event spend lands only in `investigation`, not `user_investigation`.

### PR DESCRIPTION
**Pull request by PoorneshThatturi**
# Description

Budget usage is computed per module in `llm/llm-server/budget`. The module → SQL filter map (`budget/types.go`) gave `user_investigation` (AI chat) an **empty** filter, so `getEntityTokenUsage` / `getEntityConversationCount` summed **every** conversation for the entity — including the `event-` sessions that the `investigation` module already accounts for. Result: event-analysis spend was **double-counted against the chat budget**, chat usage was inflated, and chat could return "budget exceeded" (429) prematurely.

Enforcement already partitions the two by session prefix (`api/chains.go`, `api/conversation_sync.go`): `event-` → `investigation`, otherwise → `user_investigation`. Reporting (`GetBudgetStatus`) and enforcement (`CheckBudgetLimits`) share `getEntityTokenUsage`, so the inflated number drove the premature 429s too.

**Fix:** make `user_investigation` the complement of `investigation`:

```go
ModuleUserInvestigation: " AND c.session_id NOT LIKE '" + events.SessionIdPrefixEvent + "%'",
```

`event-rca-` also starts with `event-`, so it is excluded as well. Reported usage now matches the enforced partition.

**Scope note:** the `lifecycle_storage` CTE in `getEntityTokenUsage` (provider-side cache storage cost) is intentionally left unpartitioned — per migration `V727`, account/tenant/global-scoped cache rows have `conversation_id = NULL`, so that cost isn't attributable to a single session/module. The documented double-count (conversation count + per-call token cost) is what this PR fixes.

Fixes #290

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [x] New regression tests in `budget/service_test.go`: `TestModuleQueryFilters_EventChatPartition` (filter map is a complementary event/non-event partition) and `TestGetEntityConversationCount_ModulePartition` (sqlmock-backed: the `user_investigation` count query carries `NOT LIKE 'event-%'`, `investigation` carries `LIKE 'event-%'`). Both **fail on the pre-fix code** and pass after the fix.
- [x] `go test -race ./budget/` (new tests + existing suite) passes.
- [x] `gofmt` / `go vet` clean; `golangci-lint` at the CI-pinned `v2.11.3` reports **0 issues**.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: llm/llm-server/budget/service_test.go [hunk 1]
```diff
@@ -388,3 +393,60 @@ func TestGetSystemDefaults(t *testing.T) {
...
+		require.NoError(t, err)
+		t.Cleanup(func() { _ = db.Close() })
+		return &common.DatabaseManager{Db: sqlx.NewDb(db, "postgresql")}, mock
+	}
+
+	t.Run("user_investigation excludes event sessions", func(t *testing.T) {
+		dm, mock := newDM(t)
+		// Pre-fix the filter was empty, so this NOT LIKE clause was absent and the
+		// query went unmatched — making this assertion fail.
+		mock.ExpectQuery("c.session_id NOT LIKE 'event-%'").
```
[the remark this answers is not in the document]
Comment (PoorneshThatturi) [reply llm/llm-server/budget/service_test.go]: Done in d9a40b5 — switched to `events.SessionIdPrefixEvent` instead of the hardcoded prefix. Thanks for the catch!

File: llm/llm-server/budget/service_test.go [hunk 2]
```diff
@@ -388,3 +393,60 @@ func TestGetSystemDefaults(t *testing.T) {
...
+		count, err := getEntityConversationCount(dm, "account", "acc-1", ModuleUserInvestigation, "month")
+
+		require.NoError(t, err)
+		assert.Equal(t, 42, count)
+		assert.NoError(t, mock.ExpectationsWereMet())
+	})
+
+	t.Run("investigation matches event sessions", func(t *testing.T) {
+		dm, mock := newDM(t)
+		mock.ExpectQuery("c.session_id LIKE 'event-%'").
```
[the remark this answers is not in the document]
Comment (PoorneshThatturi) [reply llm/llm-server/budget/service_test.go]: Done in d9a40b5 — switched to `events.SessionIdPrefixEvent` instead of the hardcoded prefix. Thanks for the catch!

File: llm/llm-server/budget/service_test.go [hunk 3]
```diff
@@ -388,3 +393,60 @@ func TestGetSystemDefaults(t *testing.T) {
...
+// the actual count query: user_investigation excludes event sessions while
+// investigation matches them. Driven by sqlmock, no live DB.
+func TestGetEntityConversationCount_ModulePartition(t *testing.T) {
+	newDM := func(t *testing.T) (*common.DatabaseManager, sqlmock.Sqlmock) {
+		t.Helper()
+		db, mock, err := sqlmock.New()
+		require.NoError(t, err)
+		t.Cleanup(func() { _ = db.Close() })
+		return &common.DatabaseManager{Db: sqlx.NewDb(db, "postgresql")}, mock
+	}
```
[the remark this answers is not in the document]
Comment (PoorneshThatturi) [reply llm/llm-server/budget/service_test.go]: Done in 6b51825 — moved the inline `newDM` closure out to a package-level `newMockMetastoreDM` helper. Thanks!

**Comment by PoorneshThatturi** [comment]
@mayankpande88 Addressed gemini's review comments.

File: llm/llm-server/budget/types.go [hunk 4]
```diff
@@ -20,8 +20,17 @@ var validModules = map[string]bool{
...
+	// user_investigation (AI chat) is the complement of investigation: every
+	// conversation that is NOT an event-analysis session. This mirrors the
+	// enforcement routing in api/chains.go and api/conversation_sync.go
+	// (session_id "event-" prefix -> investigation, otherwise ->
+	// user_investigation), so reported usage and enforced limits agree.
+	// Previously this filter was empty, so user_investigation usage swept in the
+	// event-* sessions that investigation already accounts for — double-counting
+	// event-analysis spend against the chat budget. ("event-rca-" also starts
+	// with "event-", so it is excluded too.)
+	ModuleUserInvestigation: " AND c.session_id NOT LIKE '" + events.SessionIdPrefixEvent + "%'",
```
Comment (hsundar-git) [inline llm/llm-server/budget/types.go]: > ✅ **Looks Good** — this correctly makes `user_investigation` the complement of `investigation`, aligning usage accounting with the enforcement routing in `api/chains.go` / `api/conversation_sync.go` (event- → investigation, else → user_investigation). Given the intended design of two **non-overlapping** budgets, this is the right fix and resolves the premature chat-budget 429s.

low

Minor: `session_id NOT LIKE 'event-%'` evaluates to `NULL` (row excluded) for any conversation with a `NULL` session_id. In practice `llm_conversations.session_id` is always set (it's part of the upsert conflict key), so this is a non-issue today — just flagging in case session-less rows ever appear, as they'd silently drop out of the chat budget.

**Comment by hsundar-git** [comment]
## Summary of Changes

Makes the `user_investigation` (AI chat) budget filter the **complement** of the `investigation` (event-analysis) filter, so the two budgets stop overlapping. Previously `user_investigation` had an empty SQL filter, so its usage/cost computation swept in `event-` sessions that `investigation` already accounts for — inflating chat usage and causing premature `429`s on the chat budget.

### Highlights

- **Budget partition fix** (`budget/types.go`): `ModuleUserInvestigation` filter `""` → `AND c.session_id NOT LIKE 'event-%'`, mirroring the enforcement routing in `api/chains.go` / `api/conversation_sync.go`.
- **Regression tests** (`budget/service_test.go`): a filter-map partition assertion + sqlmock-backed tests proving the generated count query carries `NOT LIKE` for chat and `LIKE` for events.

### Review Summary

**Service affected:** llm-server (Go) · **Risk level:** Low

Verified against the intended design (two **non-overlapping** budgets): the fix is correct and aligns reporting with the already-disjoint enforcement path. The empty filter was an original-implementation oversight (present since the budget feature shipped), not a recent regression.

| Category | Finding | Severity |
|---|---|---|
| Correctness | `types.go` partition fix is correct; resolves the premature chat-budget 429s | > ✅ Looks Good |
| Correctness | Residual overlap: the module filter is applied only to the `per_call_cost` CTE in `getEntityTokenUsage` (`budget/usage.go:166`), **not** to `lifecycle_storage`. Provider cache-**storage** cost is still summed per-entity across both modules, so it remains shared between the two budgets. PR body acknowledges this; flagging that the "no overlap" goal is met for per-call token cost but not for storage cost. Fine as a follow-up. | medium |
| Conventions | PR title scope `llm-server` is not in `.github/semantic.yml` (allowed: `llm`, `ml`, …; `requireScope: true`). The semantic PR-title check will fail — retitle to `fix(llm): …`. | medium |
| Style | `session_id NOT LIKE 'event-%'` excludes rows with `NULL` session_id; non-issue today (session_id is always set) — see inline note. | low |
| Process | **Duplicate of #386** (same one-line fix). #370 is earlier (Jun 23 vs Jun 24) and has stronger tests → keep #370, close #386. | medium |

### Checklist
- [x] Tests cover new/changed behavior (sqlmock-backed)
- [x] No hardcoded secrets
- [x] Filter built from a constant (no SQL-injection surface)
- [ ] PR title scope conforms to `semantic.yml` (currently `llm-server` → use `llm`)
- [ ] Dedupe vs #386

---
*Automated PR Review*

**Comment by hsundar-git** [comment]
Note: #386 implemented the same fix and has been closed as a duplicate in favor of this PR (this one is earlier and has the sqlmock-backed regression tests).

File: llm/llm-server/budget/types.go (hunk 4 above)
Comment (PoorneshThatturi) [reply llm/llm-server/budget/types.go]: Thanks for the careful flag! Hardened in [`4a9dd2b`](https://github.com/PoorneshThatturi/nudgebee/commit/4a9dd2b) — the filter is now `(c.session_id IS NULL OR c.session_id NOT LIKE 'event-%')`, so a session-less conversation stays in the chat budget instead of being dropped by NOT LIKE's NULL semantics (matching enforcement's otherwise -> user_investigation). Agreed it's always set today; this is just defensive.

**Comment by PoorneshThatturi** [comment]
@hsundar-git thanks for the review! All feedback is addressed — the NULL-session edge case is now guarded (`session_id IS NULL OR NOT LIKE 'event-%'`), Gemini's notes (package-level test helper, `SessionIdPrefixEvent`) are in, and the title scope is fixed. All checks are green (Go Lint, gitleaks, CLA) and it's mergeable — ready for another look / merge whenever you have a moment.

**Review by hsundar-git — APPROVED** [review]
